kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java

```diff
@@ -0,0 +1,119 @@
+package com.provectus.kafka.ui.serde.schemaregistry;
+
+
+import com.google.common.annotations.VisibleForTesting;
+import com.provectus.kafka.ui.client.BufSchemaRegistryClient;
+import com.provectus.kafka.ui.model.KafkaCluster;
+import com.provectus.kafka.ui.model.MessageSchemaDTO;
+import com.provectus.kafka.ui.model.TopicMessageSchemaDTO;
+import com.provectus.kafka.ui.serde.RecordSerDe;
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.utils.Bytes;
+
+@Slf4j
+public class BufAndSchemaRegistryAwareRecordSerDe implements RecordSerDe {
+
+  class ProtoSchema {
+    String fullyQualifiedTypeName;
+    String schemaID;
+  }
+
+  private final SchemaRegistryAwareRecordSerDe schemaRegistryAwareRecordSerDe;
+  private final KafkaCluster cluster;
+  private final BufSchemaRegistryClient bufClient;
+
+  public BufAndSchemaRegistryAwareRecordSerDe(KafkaCluster cluster) {
+    this(cluster, createBufRegistryClient(cluster));
+  }
+
+  @VisibleForTesting
+  public BufAndSchemaRegistryAwareRecordSerDe(KafkaCluster cluster, BufSchemaRegistryClient bufClient) {
+    this.schemaRegistryAwareRecordSerDe = new SchemaRegistryAwareRecordSerDe(cluster);
+    this.cluster = cluster;
+    this.bufClient = bufClient;
+  }
+
+  private static BufSchemaRegistryClient createBufRegistryClient(KafkaCluster cluster) {
+    // TODO: pass args from cluster to buf constructor
+    return new BufSchemaRegistryClient();
+  }
+
+  public DeserializedKeyValue deserialize(ConsumerRecord<Bytes, Bytes> msg) {
+    ProtoSchema protoSchema = protoSchemaFromHeaders(msg.headers());
+    if (protoSchema != null) {
+      // TODO: parse message
+      log.info("Skipping buf for schema: {}", protoSchema.fullyQualifiedTypeName);
+      return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+    }
+
+    protoSchema = protoSchemaFromTopic(msg.topic());
+
+    if (protoSchema != null) {
+      // TODO: parse message and return it
+      log.info("Skipping buf for schema: {}", protoSchema.fullyQualifiedTypeName);
+      return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+    }
+
+    return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+  }
+
+  private @Nullable ProtoSchema protoSchemaFromHeaders(Headers headers) {
```

Comment:
We should probably add units test around this (maybe just scaffold it and have it pass for now). Is it easy to add with how kafkaui is set up?

Reply:
Ah, I meant to. I'll do that now